README.md
# PolyGoneNMS

PolyGoneNMS is a library for efficient and distributed polygon Non-Maximum Suppression (NMS) in Python. It supports various NMS methods, intersection calculations, and can handle large numbers of polygons in 1D, 2D, and 3D spaces. PolyGoneNMS uses R-tree data structures and shapely polygon objects for optimal performance.

## Benchmark Plots

![Benchmark results](assets/benchmark_results.png)

## Features

- Efficient polygon NMS for large numbers of polygons.
- Support for various NMS methods: Default, Soft, and Class Agnostic.
- Support for different intersection methods: IOU, IOS, and Dice.
- R-tree data structure for efficient spatial indexing and querying.
- Distributed processing support using Ray and Dask.
- Comprehensive documentation and examples.

## Installation

You can install PolyGoneNMS using pip:

```bash
pip install polygone-nms
```

## Quickstart

```python
import numpy as np
from polygone_nms import polygone_nms

# Example input data
data = np.array([
    [0, 0, 1, 1, 0, 1, 0, 0, 1, 0.9],
    [0.5, 0.5, 1.5, 1.5, 0.5, 1.5, 0, 0, 1, 0.8],
])

# Apply NMS
results = nms(data, distributed=None, nms_method="Default", intersection_method="IOU")

print("Filtered indices:", results)

# Filtered data
print("Filtered data:")
print(data[results])
```

For a more detailed guide on using PolyGoneNMS, please see the [Quickstart](https://wolodjaz.github.io/PolyGoneNMS/0.1.6/quickstart/) in the documentation.

## Documentation

Detailed documentation is available at:
[![Docs](https://img.shields.io/badge/Docs-mkdocs-blue?style=flat)](https://wolodjaz.github.io/PolyGoneNMS)


## Contributing

We welcome contributions to the project! Please follow the usual GitHub process for submitting issues or pull requests.

## License

This project is licensed under the [MIT License](https://fossa.com/blog/open-source-licenses-101-mit-license/).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.1.7]

### Added

- updated docs with correct import of NMS algorithm.
- updated `README.md` with correct import of NMS algorithm.
- reimplemented `utils` to handle `ZeroDivisionError` errors.

## [0.1.6]

### Added

- Working Pypi upload structure

## [0.1.5]

PyPi tests
## [0.1.4]

PyPi tests
## [0.1.3]

### Added

- Pypi Publish Github Actions modified

## [0.1.2]

### Added

- CI/CD adjustments for the project in github actions.
- docs references repair.
- docs doc-strings examples corrected
- docs creation github action repair.

## [0.1.1]

### Added

- PolyGoneNMS library for efficient and distributed polygon Non-Maximum Suppression (NMS).
- Support for different NMS methods: Default, Soft, and Class Agnostic.
- Support for various intersection methods: IOU, IOS, and Dice.
- R-tree data structure for efficient spatial indexing and querying.
- Distributed processing support using Ray and Dask.
- Comprehensive documentation and examples.
- Unit and integration tests for the library.
- Pre-commit hooks, GitHub Actions, and other CI/CD configurations.


tests/test_nms.py
import functools

import numpy as np
import pytest
from shapely.geometry import Polygon

from polygone_nms.nms import (
    apply_distributed_polygon_nms,
    apply_polygon_nms,
    cluster_polygons,
    nms,
)
from polygone_nms.utils import build_rtree, iou


def test_cluster_polygons():
    polygons = [
        (Polygon([(0, 0), (1, 0), (1, 1), (0, 1)]), 0, 0.9),
        (Polygon([(2, 0), (3, 0), (3, 1), (2, 1)]), 0, 0.9),
        (Polygon([(4, 0), (5, 0), (5, 1), (4, 1)]), 0, 0.9),
        (Polygon([(0, 0.5), (1, 0.5), (1, 2), (0, 2)]), 0, 0.9),
        (Polygon([(2, 1), (3, 1), (3, 3), (2, 3)]), 0, 0.9),
    ]

    rtree = build_rtree(polygons)

    # Test cluster_polygons function
    clustered_polygons = cluster_polygons(polygons, rtree)
    assert len(clustered_polygons) == 3
    assert sorted(clustered_polygons[0]) == [0, 3]
    assert sorted(clustered_polygons[1]) == [1, 4]
    assert sorted(clustered_polygons[2]) == [2]


@pytest.mark.parametrize(
    "polygons,nms_method,intersection_method,threshold,sigma,expected",
    [
        (
            [
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
                (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
                (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
            ],
            "Default",
            iou,
            0.3,
            0.5,
            [0, 2],
        ),
        (
            [
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
                (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
                (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
            ],
            "Soft",
            iou,
            0.5,
            0.5,
            [0, 1, 2],
        ),
        (
            [
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
                (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
                (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
            ],
            "Soft",
            iou,
            0.7,
            0.5,
            [0, 2],
        ),
        (
            [
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
                (Polygon([(1, 1), (3, 1), (3, 3), (1, 3)]), 2, 0.8),
                (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
            ],
            "Class Agnostic",
            iou,
            0.5,
            0.5,
            [0, 1, 2],
        ),
        (
            [
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.7),
            ],
            "Default",
            iou,
            0.5,
            0.5,
            [0],
        ),
        (
            [
                (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
                (Polygon([(1, 1), (3, 1), (3, 3), (1, 3)]), 1, 0.8),
                (Polygon([(2, 2), (4, 2), (4, 4), (2, 4)]), 2, 0.85),
                (Polygon([(3, 3), (5, 3), (5, 5), (3, 5)]), 1, 0.7),
                (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 2, 0.95),
                (Polygon([(5, 5), (7, 5), (7, 7), (5, 7)]), 1, 0.6),
            ],
            "Class Agnostic",
            iou,
            0.12,
            0.5,
            [0, 2, 4],
        ),
    ],
)
def test_apply_polygon_nms(
    polygons, nms_method, intersection_method, threshold, sigma, expected
):
    result = apply_polygon_nms(
        polygons, nms_method, intersection_method, threshold, sigma
    )
    assert sorted(result) == expected


def test_fail_apply_polygon_nms():
    polygons = [
        (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
        (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
        (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
    ]
    with pytest.raises(ValueError) as excinfo:
        apply_polygon_nms(polygons, "Unknown", iou, 0.3, 0.5)
    assert "Invalid NMS method: Unknown." in str(excinfo.value)


@pytest.mark.ray_distributed
def test_ray_polygon_nms():
    import ray

    ray.init()

    polygons_first = [
        (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
        (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
        (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
    ]
    polygons_second = [
        (Polygon([(10, 10), (12, 10), (12, 12), (10, 12)]), 1, 0.9),
        (Polygon([(11, 10), (13, 10), (13, 12), (11, 12)]), 1, 0.8),
        (Polygon([(14, 14), (16, 14), (16, 16), (14, 16)]), 1, 0.95),
    ]
    clusters = [polygons_first, polygons_second]
    nms_polygons_futures = [
        ray.remote(apply_distributed_polygon_nms).remote(
            cluster_polygons, "Default", iou, 0.3, 0.5
        )
        for cluster_polygons in clusters
    ]
    nms_polygons = [
        nms_polygon
        for nms_cluster in ray.get(nms_polygons_futures)
        for nms_polygon in nms_cluster
    ]

    assert sorted(nms_polygons) == [0, 0, 2, 2]
    ray.shutdown()


@pytest.mark.dask_distributed
def test_dask_polygon_nms():
    from dask.distributed import Client

    polygons_first = [
        (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
        (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
        (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
    ]
    polygons_second = [
        (Polygon([(10, 10), (12, 10), (12, 12), (10, 12)]), 1, 0.9),
        (Polygon([(11, 10), (13, 10), (13, 12), (11, 12)]), 1, 0.8),
        (Polygon([(14, 14), (16, 14), (16, 16), (14, 16)]), 1, 0.95),
    ]
    clusters = [polygons_first, polygons_second]

    with Client() as client:
        nms_polygon_futures = client.map(
            functools.partial(
                apply_distributed_polygon_nms,
                nms_method="Default",
                intersection_method=iou,
                threshold=0.3,
                sigma=0.5,
            ),
            clusters,
        )
        results = client.gather(nms_polygon_futures)
        nms_polygons = [idx for result in results for idx in result]

    assert sorted(nms_polygons) == [0, 0, 2, 2]


@pytest.mark.parametrize(
    "nms_method,intersection_method,threshold,sigma,expected",
    [
        ("Default", "IOU", 0.3, 0.5, [0, 2, 3, 5]),
        ("Default", "IOS", 0.3, 0.5, [0, 2, 3, 5]),
        ("Default", "Dice", 0.3, 0.5, [0, 2, 3, 5]),
        ("Default", "IOT", 0.3, 0.5, [0, 2, 3, 5]),
        ("Soft", "IOU", 0.7, 0.5, [0, 2, 3, 5]),
        ("Class Agnostic", "IOU", 0.3, 0.5, [0, 2, 3, 5]),
    ],
)
def test_nms(nms_method, intersection_method, threshold, sigma, expected):
    input_data = np.array(
        [
            [0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9],
            [1.0, 0.0, 3.0, 0.0, 3.0, 2.0, 1.0, 2.0, 1.0, 0.8],
            [4.0, 4.0, 6.0, 4.0, 6.0, 6.0, 4.0, 6.0, 5.0, 0.95],
            [10.0, 10.0, 12.0, 10.0, 12.0, 12.0, 10.0, 12.0, 11.0, 0.9],
            [11.0, 10.0, 13.0, 10.0, 13.0, 12.0, 11.0, 12.0, 11.0, 0.8],
            [14.0, 14.0, 16.0, 14.0, 16.0, 16.0, 14.0, 16.0, 15.0, 0.95],
        ]
    )

    results = nms(input_data, None, nms_method, intersection_method, threshold, sigma)
    assert sorted(results) == expected


def test_nms_with_different_input_types():
    input_data = [
        np.array([0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9]),
        np.array([1.0, 0.0, 3.0, 0.0, 3.0, 2.0, 1.0, 2.0, 1.0, 0.8]),
        np.array([4.0, 4.0, 6.0, 4.0, 6.0, 6.0, 4.0, 6.0, 5.0, 0.95]),
        np.array([10.0, 10.0, 12.0, 10.0, 12.0, 12.0, 10.0, 12.0, 11.0, 0.9]),
        np.array([11.0, 10.0, 13.0, 10.0, 13.0, 12.0, 11.0, 12.0, 11.0, 0.8]),
        np.array([14.0, 14.0, 16.0, 14.0, 16.0, 16.0, 14.0, 16.0, 15.0, 0.95]),
    ]

    results = nms(input_data, None, "Default", "IOU", 0.3, 0.5)
    assert sorted(results) == [0, 2, 3, 5]

    input_data_polygons = [
        (Polygon([(0, 0), (2, 0), (2, 2), (0, 2)]), 1, 0.9),
        (Polygon([(1, 0), (3, 0), (3, 2), (1, 2)]), 1, 0.8),
        (Polygon([(4, 4), (6, 4), (6, 6), (4, 6)]), 1, 0.95),
        (Polygon([(10, 10), (12, 10), (12, 12), (10, 12)]), 1, 0.9),
        (Polygon([(11, 10), (13, 10), (13, 12), (11, 12)]), 1, 0.8),
        (Polygon([(14, 14), (16, 14), (16, 16), (14, 16)]), 1, 0.95),
    ]

    results = nms(input_data_polygons, None, "Default", "IOU", 0.3, 0.5)
    assert sorted(results) == [0, 2, 3, 5]

    input_empty_list = []
    results = nms(input_empty_list, None, "Default", "IOU", 0.3, 0.5)
    assert results == []


def test_fail_nms():
    dump_data = np.array([[0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9]])

    with pytest.raises(ValueError) as excinfo:
        nms(0, None, "Default", "IOU", 0.3, 0.5)
    assert (
        "Invalid input data type. Expected a list of polygons or a 2D NumPy array."
        in str(excinfo.value)
    )

    with pytest.raises(ValueError) as excinfo:
        nms(dump_data, "Wrong", "Default", "IOU", 0.3, 0.5)

    assert "Invalid distributed computing method: Wrong." in str(excinfo.value)

    with pytest.raises(ValueError) as excinfo:
        nms(dump_data, None, "Wrong", "IOU", 0.3, 0.5)

    assert "Invalid NMS method: Wrong." in str(excinfo.value)

    with pytest.raises(ValueError) as excinfo:
        nms(dump_data, None, "Default", "Wrong", 0.3, 0.5)

    assert "Invalid intersection method: Wrong." in str(excinfo.value)


@pytest.mark.ray_distributed
def test_ray_nms():
    import ray

    input_data = np.array(
        [
            [0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9],
            [1.0, 0.0, 3.0, 0.0, 3.0, 2.0, 1.0, 2.0, 1.0, 0.8],
            [4.0, 4.0, 6.0, 4.0, 6.0, 6.0, 4.0, 6.0, 5.0, 0.95],
            [10.0, 10.0, 12.0, 10.0, 12.0, 12.0, 10.0, 12.0, 11.0, 0.9],
            [11.0, 10.0, 13.0, 10.0, 13.0, 12.0, 11.0, 12.0, 11.0, 0.8],
            [14.0, 14.0, 16.0, 14.0, 16.0, 16.0, 14.0, 16.0, 15.0, 0.95],
        ]
    )

    results = nms(input_data, "Ray", "Default", "IOU", 0.3, 0.5)
    assert sorted(results) == [0, 2, 3, 5]
    assert not ray.is_initialized()

    ray.init()
    assert ray.is_initialized()
    results = nms(input_data, "Ray", "Default", "IOU", 0.3, 0.5)
    assert sorted(results) == [0, 2, 3, 5]
    assert ray.is_initialized()
    ray.shutdown()
    assert not ray.is_initialized()


@pytest.mark.dask_distributed
def test_dask_nms():
    import dask
    from dask.distributed import Client

    input_data = np.array(
        [
            [0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9],
            [1.0, 0.0, 3.0, 0.0, 3.0, 2.0, 1.0, 2.0, 1.0, 0.8],
            [4.0, 4.0, 6.0, 4.0, 6.0, 6.0, 4.0, 6.0, 5.0, 0.95],
            [10.0, 10.0, 12.0, 10.0, 12.0, 12.0, 10.0, 12.0, 11.0, 0.9],
            [11.0, 10.0, 13.0, 10.0, 13.0, 12.0, 11.0, 12.0, 11.0, 0.8],
            [14.0, 14.0, 16.0, 14.0, 16.0, 16.0, 14.0, 16.0, 15.0, 0.95],
        ]
    )

    results = nms(input_data, "Dask", "Default", "IOU", 0.3, 0.5)
    assert sorted(results) == [0, 2, 3, 5]
    assert not dask.config.get("distributed", {}).get("active", False)

    with Client() as client:
        results = nms(input_data, "Dask", "Default", "IOU", 0.3, 0.5, client=client)
        assert sorted(results) == [0, 2, 3, 5]
    assert not dask.config.get("distributed", {}).get("active", False)


tests/test_polygone_nms.py
import os

import numpy as np

import polygone_nms


def get_version():
    with open(os.path.join(os.path.dirname(__file__), "..", "VERSION")) as f:
        return f.read().strip()


def test_polygone_nms():
    assert polygone_nms.__version__ == get_version()


def test_polygone_nms_nms():
    input_data = np.array(
        [
            [0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9],
            [1.0, 0.0, 3.0, 0.0, 3.0, 2.0, 1.0, 2.0, 1.0, 0.8],
            [4.0, 4.0, 6.0, 4.0, 6.0, 6.0, 4.0, 6.0, 5.0, 0.95],
            [10.0, 10.0, 12.0, 10.0, 12.0, 12.0, 10.0, 12.0, 11.0, 0.9],
            [11.0, 10.0, 13.0, 10.0, 13.0, 12.0, 11.0, 12.0, 11.0, 0.8],
            [14.0, 14.0, 16.0, 14.0, 16.0, 16.0, 14.0, 16.0, 15.0, 0.95],
        ]
    )

    results = polygone_nms.nms(input_data, None, "Default", "IOU", 0.3, 0.5)
    assert sorted(results) == [0, 2, 3, 5]


scripts/check_version.py
import subprocess
import sys


def main():
    version_changed = False
    benchmark_changed = False

    output = subprocess.check_output(["git", "diff", "--cached", "--name-only"]).decode(
        "utf-8"
    )

    if "VERSION" in output:
        version_changed = True

    if "assets/benchmark_results.png" in output:
        benchmark_changed = True

    if version_changed:
        print("REMEMBER: Add changes to CHANGELOG.md")
    if version_changed and not benchmark_changed:
        print(
            "ERROR: The VERSION file has changed, "
            "but the assets/benchmark_results.png file has not. "
            "Please update the benchmark results before committing."
        )
        sys.exit(1)
    elif version_changed:
        print(
            "WARNING: The VERSION file has changed. "
            "Remember to run benchmark.py if applicable."
        )


if __name__ == "__main__":
    main()


polygone_nms/nms.py
# Import necessary libraries
import functools
from typing import Callable, List, Optional, Tuple, Union

import numpy as np
import rtree
from shapely.geometry import Polygon

# Import package modules
from polygone_nms.utils import (  # , dfs_recursive
    build_rtree,
    create_polygon,
    dfs_iterative,
    dice,
    ios,
    iot,
    iou,
)

INTERSECTION_METHODS = {
    "IOU": iou,
    "IOT": iot,
    "IOS": ios,
    "Dice": dice,
}

VALID_NMS_METHODS = ["Default", "Soft", "Class Agnostic"]
VALID_DISTRIBUTED_METHODS = ["Dask", "Ray"]


def cluster_polygons(
    polygons: List[Tuple[Polygon, float, float]], rtree_index: rtree.index.Index
) -> List[List[Polygon]]:
    """
    Cluster polygons into non-overlapping subregions with R-Tree.
    Used for distributed computing.

    Examples:
        >>> from shapely.geometry import Polygon
        >>> from polygone_nms.utils import build_rtree, cluster_polygons
        >>> polygons = [
        ...     (Polygon([(0, 0), (1, 0), (1, 1), (0, 1)]), 0, 0.9),
        ...     (Polygon([(2, 0), (3, 0), (3, 1), (2, 1)]), 0, 0.9),
        ...     (Polygon([(4, 0), (5, 0), (5, 1), (4, 1)]), 0, 0.9),
        ...     (Polygon([(0, 0.5), (1, 0.5), (1, 2), (0, 2)]), 0, 0.9),
        ...     (Polygon([(2, 1), (3, 1), (3, 3), (2, 3)]), 0, 0.9),
        ... ]
        >>> rtree = build_rtree(polygons)
        >>> clustered_polygons = cluster_polygons(polygons, rtree)
        >>> assert len(clustered_polygons) == 3
        >>> assert sorted(clustered_polygons[0]) == [0, 3]
        >>> assert sorted(clustered_polygons[1]) == [1, 4]

    Args:
        polygons (List[Tuple[Polygon, float, float]]): List of polygons,
            where each polygon is a tuple of the polygon, the class label,
            and the confidence score.
        rtree_index (rtree.index.Index): R-Tree index of the polygons.

    Returns:
        List[List[Polygon]]:
            A list of clusters,
            where each cluster is a list of non-overlapping polygons.
    """
    # Create a list of adjacent polygons for each polygon and
    # a list of visited polygons
    adj_list: List[List[int]] = [[] for _ in range(len(polygons))]
    visited = [False] * len(polygons)

    # Iterate over all polygons
    for i, poly_tuple in enumerate(polygons):
        # Get the polygon
        poly = poly_tuple[0]
        # Get all polygons that intersect with the current polygon
        intersecting_polygons = list(rtree_index.intersection(poly.bounds))
        # Remove the current polygon from the list of intersecting polygons
        intersecting_polygons.remove(i)
        # Update the adjacency list
        adj_list[i] = intersecting_polygons

    # Create a list of clusters
    clusters = []
    # Iterate over all polygons
    for i in range(len(polygons)):
        # If the current polygon has not been visited
        if not visited[i]:
            # Perform a depth-first search to find all connected components
            # Due to RecursionError: maximum recursion depth exceeded
            # connected_component = dfs_recursive(i, visited, adj_list)
            connected_component = dfs_iterative(i, visited, adj_list)
            # # Create a cluster from the connected component
            # cluster = [polygons[j] for j in connected_component]
            # # Append the cluster to the list of clusters
            # Append the connected component (cluster) to the list of clusters
            clusters.append(connected_component)

    return clusters


def apply_polygon_nms(
    polygons: List[Tuple[Polygon, float, float]],
    nms_method: str,
    intersection_method: Callable,
    threshold: float = 0.5,
    sigma: float = 0.5,
) -> List[int]:
    """
    Apply Non-Maximum Suppression (NMS) to a list of predicted polygons.

    Raises:
        ValueError: If the NMS method is invalid.

    Args:
        polygons (List[Tuple[Polygon, float, float]]):
            List of polygons, where each polygon is a tuple of the polygon
            represented by shapely.geometry.Polygon object, the class label,
            and the confidence score.
        nms_method (str): The NMS method to use, one of
            ("Default", "Soft", "Class Agnostic").
        intersection_method (Callable): The method to compute intersections.
        threshold (float, optional): The threshold for the NMS method. Defaults to 0.5.
        sigma (float, optional): The sigma for the Soft NMS method. Defaults to 0.5.

    Returns:
        List[int]: A list of kept polygon indices.
    """
    # List of kept polygons
    kept_polygons = []

    # Sort polygons by confidence score
    confidences = np.array([poly[2] for poly in polygons])
    sorted_indices = np.argsort(confidences)[::-1]

    # Iterate over all sorted polygons by confidence score
    while sorted_indices.size > 0:
        # Get the index of the current polygon and the respective polygon and label
        i = sorted_indices[0]
        current_polygon = polygons[i][0]
        current_class = polygons[i][1]

        # Add the current polygon to the NMS kept polygons list
        kept_polygons.append(i)

        # Calculate the intersection between the current polygon and
        # all remaining polygons
        intersections = [
            intersection_method(current_polygon, polygons[j][0])
            for j in sorted_indices[1:]
        ]

        # If the NMS method is "Default", "Soft" or "Class Agnostic"
        if nms_method == "Default":
            # Remove all remaining polygons that have an intersection with
            # the current polygon greater than the threshold
            remaining_indices = [
                idx
                for idx, intersection in zip(sorted_indices[1:], intersections)
                if polygons[idx][1] != current_class or intersection <= threshold
            ]
        elif nms_method == "Soft":
            # Update the confidence scores of the remaining polygons
            for j, intersection in zip(sorted_indices[1:], intersections):
                # Calculate the weight for the current polygon equal to
                # e^(-intersection^2 / sigma)
                weight = np.exp(-(intersection**2) / sigma)
                confidences[j] *= weight

            # Discard polygons with confidence below the threshold
            remaining_indices = [
                idx for idx in sorted_indices[1:] if confidences[idx] >= threshold
            ]
        elif nms_method == "Class Agnostic":
            # Remove all remaining polygons that have an intersection with
            # the current polygon greater than the threshold and
            # have the same class label
            remaining_indices = [
                idx
                for idx, intersection in zip(sorted_indices[1:], intersections)
                if intersection <= threshold
            ]
        else:
            raise ValueError(
                (
                    f"Invalid NMS method: {nms_method}. "
                    f"Allowed methods are: {VALID_NMS_METHODS}"
                )
            )

        # Update the list of remaining polygon indices
        sorted_indices = np.array(remaining_indices)

    return kept_polygons


def apply_distributed_polygon_nms(
    polygons: List[Tuple[Polygon, float, float]],
    nms_method: str,
    intersection_method: Callable,
    threshold: float = 0.5,
    sigma: float = 0.5,
) -> List[int]:
    """
    Distributed version of `apply_polygon_nms`.

    Args:
        polygons (List[Tuple[Polygon, float, float]]):
            List of polygons, where each polygon is a tuple of the polygon
            represented by shapely.geometry.Polygon object, the class label,
            and the confidence score.
        nms_method (str): The NMS method to use, one of
            ("Default", "Soft", "Class Agnostic").
        intersection_method (Callable): The method to compute intersections.
        threshold (float, optional): The threshold for the NMS method. Defaults to 0.5.
        sigma (float, optional): The sigma for the Soft NMS method. Defaults to 0.5.

    Returns:
        List[int]: A list of kept polygon indices.
    """
    return apply_polygon_nms(
        polygons, nms_method, intersection_method, threshold, sigma
    )


def nms(
    input_data: Union[np.ndarray, Tuple[Polygon, float, float]],
    distributed: Optional[str] = None,
    nms_method: str = "Default",
    intersection_method: str = "IOU",
    threshold: float = 0.5,
    sigma: float = 0.5,
    **kwargs,
) -> List[int]:
    """
    Apply Non-Maximum Suppression (NMS) to a set of polygons.
    Method works with distributed computing for efficient processing and clustering.

    Examples:
        >>> import numpy as np
        >>> from polygone.nms import nms
        >>> input_data = np.array(
        ... [
        ...     [0.0, 0.0, 2.0, 0.0, 2.0, 2.0, 0.0, 2.0, 1.0, 0.9],
        ...     [1.0, 0.0, 3.0, 0.0, 3.0, 2.0, 1.0, 2.0, 1.0, 0.8],
        ...     [4.0, 4.0, 6.0, 4.0, 6.0, 6.0, 4.0, 6.0, 5.0, 0.95],
        ...     [10.0, 10.0, 12.0, 10.0, 12.0, 12.0, 10.0, 12.0, 11.0, 0.9],
        ...     [11.0, 10.0, 13.0, 10.0, 13.0, 12.0, 11.0, 12.0, 11.0, 0.8],
        ...     [14.0, 14.0, 16.0, 14.0, 16.0, 16.0, 14.0, 16.0, 15.0, 0.95],
        ... ])
        >>> results = nms(input_data, None, "Default", "IOU", 0.3, 0.5)
        >>> assert sorted(results) == [0, 2, 3, 5]

    results = nms(input_data, None, nms_method, intersection_method, threshold, sigma)
    assert sorted(results) == expected

    Args:
        input_data (Union[np.ndarray, Tuple[Polygon, float, float]]):
            List of tuples of Polygon, class, score or numpy array of polygons.
            Each polygon in the numpy array is represented by a 1D array of n % 2
            coordinates (x1, y1, x2, y2, .., x(n-1), y(n-1), class, score).
        distributed (Optional[str], optional):
            The distributed computing method to use,
            one of (None, "Ray", "Dask").. Defaults to None.
        nms_method (str, optional):
            The NMS method to use, one of ("Default", "Soft", "Class Agnostic").
            Defaults to "Default".
        intersection_method (str, optional):
            The method to compute intersections, one of ("IOU", "IOS", "Dice", "IOT").
            Defaults to "IOU".
        threshold (float, optional):
            The threshold for the NMS(intersection) method. Defaults to 0.5.
        sigma (float, optional):
            The sigma for the Soft NMS method. Defaults to 0.5.
        **kwargs: Additional arguments for the NMS method.
            Any keyword arguments for the distributed computing should be passed here.

    Returns:
        List[int]: List of indices of the kept polygons.
    """
    # Check if the input data is a list of polygons or a 2D NumPy array
    polygons = []
    if isinstance(input_data, list):
        if len(input_data) == 0:
            return []
        elif isinstance(input_data[0], tuple):
            polygons = input_data
        else:
            input_data = np.array(input_data)
    elif not isinstance(input_data, np.ndarray):
        raise ValueError(
            "Invalid input data type. Expected a list of polygons or a 2D NumPy array."
        )

    # Check distributed computing method
    if distributed is not None:
        if distributed not in VALID_DISTRIBUTED_METHODS:
            raise ValueError(
                (
                    f"Invalid distributed computing method: {distributed}. "
                    f"Allowed methods are: {VALID_DISTRIBUTED_METHODS}"
                )
            )

    # Check NMS method
    if nms_method not in VALID_NMS_METHODS:
        raise ValueError(
            (
                f"Invalid NMS method: {nms_method}. "
                f"Allowed methods are: {VALID_NMS_METHODS}"
            )
        )

    # Check intersection method
    if intersection_method not in list(INTERSECTION_METHODS.keys()):
        raise ValueError(
            (
                f"Invalid intersection method: {intersection_method}. "
                f"Allowed methods are: {list(INTERSECTION_METHODS.keys())}"
            )
        )

    # Get the intersection function
    intersection_func = INTERSECTION_METHODS[intersection_method]

    # Convert input data to Shapely Polygons and store class and confidence
    if len(polygons) == 0:
        for row in input_data:
            polygon_coords = row[:-2]
            class_label = row[-2]
            confidence = row[-1]
            polygon = create_polygon(polygon_coords)
            polygons.append((polygon, class_label, confidence))

    # Build R-Tree index
    rtree = build_rtree(polygons)

    # Cluster polygons into non-overlapping subregions
    clusters = cluster_polygons(polygons, rtree)
    polygon_clusters = [[polygons[idx] for idx in cluster] for cluster in clusters]

    # Apply NMS to the clustered polygons
    if distributed == "Ray":
        # Import Ray
        import ray

        # Check if Ray is initialized
        if ray.is_initialized():
            initialize_ray = False
        else:
            initialize_ray = True
            ray.init()  # Initialize Ray

        # Scatter the polygon clusters to the Ray cluster
        polygon_futures = [
            ray.put(polygon_cluster) for polygon_cluster in polygon_clusters
        ]

        # Apply NMS to each cluster
        nms_polygons_futures = [
            ray.remote(apply_distributed_polygon_nms).remote(
                polygon_future, nms_method, intersection_func, threshold, sigma
            )
            for polygon_future in polygon_futures
        ]
        nms_polygons = ray.get(nms_polygons_futures)

        # Shutdown Ray if it was initialized in this function
        if initialize_ray:
            ray.shutdown()  # Shutdown Ray
    elif distributed == "Dask":
        # Use an existing Dask client if provided, or create a new one
        client = kwargs.get("client", None)
        if client is None:
            from dask.distributed import Client

            initialize_dask = True
            client = Client()
        else:
            initialize_dask = False

        # Scatter the polygon clusters to the Dask cluster
        polygon_futures = client.scatter(polygon_clusters)

        # Apply NMS to each cluster and gather the results
        nms_polygon_futures = client.map(
            functools.partial(
                apply_distributed_polygon_nms,
                nms_method=nms_method,
                intersection_method=intersection_func,
                threshold=threshold,
                sigma=sigma,
            ),
            polygon_futures,
        )
        nms_polygons = client.gather(nms_polygon_futures)
        # Shutdown Dask if it was initialized in this function
        if initialize_dask:
            client.close()  # Shutdown Dask
    else:
        # Apply NMS to each cluster
        nms_polygons = [
            apply_distributed_polygon_nms(
                polygon_cluster, nms_method, intersection_func, threshold, sigma
            )
            for polygon_cluster in polygon_clusters
        ]

    # Combine the kept polygon indices from each cluster and
    # get the keep indices from each cluster
    keep_indices = []
    for cluster, cluster_nms in zip(clusters, nms_polygons):
        for idx in cluster_nms:
            keep_indices.append(cluster[idx])

    return keep_indices
